sol/main.c

```diff
@@ -0,0 +1,507 @@
+		goto drop;
+	}
+
+	next_lowest_pr = list_last_entry(&lowest_pr->list,
+		struct priority_req, list);
+
+	/* The lowest priority packet was the only one of that priority. */
+	if (lowest_pr->priority != next_lowest_pr->priority) {
+		req_queue->priorities[lowest_pr->priority] = NULL;
+		req_queue->lowest_priority = next_lowest_pr->priority;
+		goto drop;
+	}
+
+	req_queue->priorities[lowest_pr->priority] = next_lowest_pr;
+
+drop:
+	list_del(&lowest_pr->list);
+	mb_free_entry(&sol_conf->mb, lowest_pr);
+	rte_pktmbuf_free(lowest_pr_pkt);
+	req_queue->len--;
+}
+
+static void
+enqueue_req(struct sol_config *sol_conf, struct priority_req *req)
+{
+	struct req_queue *req_queue = &sol_conf->req_queue;
+	uint8_t priority = req->priority;
+
+	if (req_queue->len == sol_conf->pri_req_max_len) {
+		/* New packet is lowest priority, so drop it. */
+		if (req_queue->lowest_priority >= priority) {
+			struct rte_mbuf *lowest_pri_pkt = req->pkt;
+			mb_free_entry(&sol_conf->mb, req);
+			rte_pktmbuf_free(lowest_pri_pkt);
+			return;
+		}
+		drop_lowest_priority_pkt(sol_conf, req_queue);
+	}
+
+	if (req_queue->priorities[priority] == NULL) {
+		/* Insert request of a priority we don't yet have. */
+		insert_new_priority_req(req_queue, req, priority);
+	} else {
+		/* Append request to end of the appropriate priority. */
+		list_add(&req->list, &req_queue->priorities[priority]->list);
+		req_queue->priorities[priority] = req;
+	}
+
+	req_queue->len++;
+}
+
+static void
+enqueue_reqs(struct sol_config *sol_conf)
+{
+	struct priority_req *req_nodes[sol_conf->enq_burst_size];
+	int num_reqs = mb_dequeue_burst(&sol_conf->mb,
+		(void **)req_nodes, sol_conf->enq_burst_size);
+	int i;
+	for (i = 0; i < num_reqs; i++)
+		enqueue_req(sol_conf, req_nodes[i]);
+}
+
+static inline void
+time_resync(struct req_queue *req_queue)
+{
+	uint64_t cycles = rte_get_tsc_cycles();
+	uint64_t cycles_diff = cycles - req_queue->time_cpu_cycles;
+	uint64_t bytes_diff;
+
+	/* Compute elapsed time in bytes. */
+	bytes_diff = rte_reciprocal_divide(cycles_diff << GATEKEEPER_TIME_SHIFT,
+		req_queue->inv_cycles_per_byte);
+
+	/* Advance port time. */
+	req_queue->time_cpu_cycles = cycles;
+	req_queue->time_cpu_bytes += bytes_diff;
+	if (req_queue->time_bytes < req_queue->time_cpu_bytes)
+		req_queue->time_bytes = req_queue->time_cpu_bytes;
+}
+
+static inline void
+credits_update(struct req_queue *req_queue)
+{
+	uint64_t n_periods =
+		(req_queue->time_bytes - req_queue->tb_time_bytes) /
+		req_queue->tb_period_bytes;
+	req_queue->tb_credits += n_periods * req_queue->tb_credits_per_period;
+	req_queue->tb_credits = req_queue->tb_credits < req_queue->tb_size
+		? req_queue->tb_credits
+		: req_queue->tb_size;
+	req_queue->tb_time_bytes += n_periods * req_queue->tb_period_bytes;
+}
+
+static inline int
+credits_check(struct req_queue *req_queue, struct rte_mbuf *pkt)
+{
+	uint32_t pkt_len = pkt->pkt_len + RTE_SCHED_FRAME_OVERHEAD_DEFAULT;
```

Comment:
That constant defines the Ethernet framing overhead as bytes that count toward the number of credits used. Its definition is:

    /*
     * Ethernet framing overhead. Overhead fields per Ethernet frame:
     * 1. Preamble:                             7 bytes;
     * 2. Start of Frame Delimiter (SFD):       1 byte;
     * 3. Frame Check Sequence (FCS):           4 bytes;
     * 4. Inter Frame Gap (IFG):               12 bytes.
     *
     * The FCS is considered overhead only if not included in the packet
     * length (field pkt_len of struct rte_mbuf).
     */
    #ifndef RTE_SCHED_FRAME_OVERHEAD_DEFAULT
    #define RTE_SCHED_FRAME_OVERHEAD_DEFAULT      24
    #endif

I'll add a comment for this.